datalad/core/dataset/gitrepo.py

```diff
@@ -0,0 +1,531 @@
+
+        Yields
+        ------
+        dict with items matching the given `fields`
+
+        Raises
+        ------
+        ValueError
+          if no `fields` are given
+
+        RuntimeError
+          if `git for-each-ref` returns a record where the number of
+          properties does not match the number of `fields`
+        """
+        if not fields:
+            raise ValueError('no `fields` provided, refuse to proceed')
+        fields = ensure_list(fields)
+        cmd = [
+            "for-each-ref",
+            "--format={}".format(
+                '%00'.join(
+                    '%({})'.format(f) for f in fields)),
+        ]
+        if points_at:
+            cmd.append('--points-at={}'.format(points_at))
+        if contains:
+            cmd.append('--contains={}'.format(contains))
+        if sort:
+            for k in ensure_list(sort):
+                cmd.append('--sort={}'.format(k))
+        if pattern:
+            cmd += ensure_list(pattern)
+        if count:
+            cmd.append('--count={:d}'.format(count))
+
+        out, _ = self._call_git(cmd)
+        for line in out.splitlines():
+            props = line.split('\0')
+            if len(fields) != len(props):
+                raise RuntimeError(
+                    'expected fields {} from git-for-each-ref, but got: {}'.format(
+                        fields, props))
+            yield dict(zip(fields, props))
+
+    def _call_git(self, args, files=None, expect_stderr=False, expect_fail=False,
+                  honor_fake_dates=True, cwd=None, env=None, index_file=None):
+        """Allows for calling arbitrary commands.
+
+        Internal helper to the call_git*() methods.
+
+        Parameters
+        ----------
+        args : list of str
+          Arguments to pass to `git`.
+        files : list of str, optional
+          File arguments to pass to `git`. The advantage of passing these here
+          rather than as part of `args` is that the call will be split into
+          multiple calls to avoid exceeding the maximum command line length.
+        expect_stderr : bool, optional
+          Standard error is expected and should not be elevated above the DEBUG
+          level.
+        expect_fail : bool, optional
+          A non-zero exit is expected and should not be elevated above the
+          DEBUG level.
+        honor_fake_dates : bool, optional
+          Whether to actually fake dates for Git operations, if enable for a
+          repo.
+
+
+        Returns
+        -------
+        stdout, stderr
+
+        Raises
+        ------
+        CommandError if the call exits with a non-zero status.
+        """
+        cmd = ['git'] + self._GIT_COMMON_OPTIONS + args
+
+        if honor_fake_dates and self.fake_dates_enabled:
+            env = self.add_fake_dates(env)
+
+        if index_file:
+            env = (env if env is not None else environ).copy()
+            env['GIT_INDEX_FILE'] = index_file
```

Comment:
AFAICS `index_file` is unused.  Should `_call_git` drop the parameter?

Reply:
It should. I will try.

Reply:
Done